File: klinara-android/app/src/main/java/com/klinara/android/services/branches/BranchesService.kt
package com.klinara.android.services.branches

import com.klinara.android.services.crm.Patch
import com.klinara.android.services.mock.MockErrors
import com.klinara.android.services.mock.MockIds
import com.klinara.android.services.networking.ApiClient
import com.klinara.android.services.networking.ApiError
import com.klinara.android.services.networking.ApiRequest
import com.klinara.android.services.networking.KlinaraJson
import com.klinara.android.services.networking.ListEnvelope
import com.klinara.android.services.networking.RequestBodyPayload
import kotlinx.coroutines.delay
import kotlinx.serialization.json.JsonObject
import java.time.Instant

/**
 * Şube yönetimi uçları (A7.4) — iOS `BranchesService` paritesi.
 *
 * `GET branches` girişte `AuthService.branches()` ile de çağrılıyor ama orada dar
 * `BranchSummary` yeterli; yönetim ekranı tam şekle bakıyor. Silme ucu YOK.
 */
interface BranchesService {
    /** `GET branches` — pasifler dahil kiracının tüm şubeleri. */
    suspend fun list(): List<BranchDetail>

    /** `POST branches` — aynı kodla ikinci şube 409. */
    suspend fun create(input: CreateBranchInput): BranchDetail

    /** `PATCH branches/:id`. */
    suspend fun update(
        id: String,
        input: UpdateBranchInput,
    ): BranchDetail
}

class LiveBranchesService internal constructor(
    private val client: ApiClient,
) : BranchesService {
    override suspend fun list(): List<BranchDetail> =
        client.send<ListEnvelope<BranchDetail>>(ApiRequest.get("branches")).data

    override suspend fun create(input: CreateBranchInput): BranchDetail =
        client.send(ApiRequest.post("branches", body = input.toJson().asBody()))

    override suspend fun update(
        id: String,
        input: UpdateBranchInput,
    ): BranchDetail = client.send(ApiRequest.patch("branches/$id", body = input.toJson().asBody()))
}

private fun JsonObject.asBody(): RequestBodyPayload =
    RequestBodyPayload(KlinaraJson.encodeToString(JsonObject.serializer(), this))

/**
 * Bellek-içi şube tablosu. [MockIds] şubeleriyle tohumlu: personel ve takvim mock'larının
 * şube kimlikleri burada da tanınsın. Sunucunun "aynı kod 409" kuralını taklit eder.
 */
class MockBranchesService(
    private val latencyEnabled: Boolean = true,
    /** Ağ hatası senaryosunda şubeler de düşsün diye; container kurarken verilir. */
    var failing: Boolean = false,
) : BranchesService {
    private val records: MutableList<BranchDetail> = ALL.toMutableList()
    private var idCounter = 0

    override suspend fun list(): List<BranchDetail> {
        settle()
        return records.toList()
    }

    override suspend fun create(input: CreateBranchInput): BranchDetail {
        settle()
        if (records.any { it.slug == input.slug }) {
            throw MockErrors.conflict(
                "Bu şube kodu zaten kullanımda",
                "\"${input.slug}\" bu klinikte başka bir şubeye ait.",
            )
        }
        idCounter += 1
        val created =
            BranchDetail(
                id = "b-mock-$idCounter",
                tenantId = MockIds.TENANT_NISANTASI,
                slug = input.slug,
                name = input.name,
                timezone = input.timezone ?: "Europe/Istanbul",
                phone = input.phone,
                address = input.address,
                createdAt = Instant.now(),
            )
        records += created
        return created
    }

    override suspend fun update(
        id: String,
        input: UpdateBranchInput,
    ): BranchDetail {
        settle()
        val index = records.indexOfFirst { it.id == id }
        if (index < 0) throw MockErrors.notFound("Şube")
        val old = records[index]
        val updated =
            old.copy(
                name = input.name ?: old.name,
                timezone = input.timezone ?: old.timezone,
                phone = input.phone.resolve(old.phone),
                address = input.address.resolve(old.address),
                isActive = input.isActive ?: old.isActive,
            )
        records[index] = updated
        return updated
    }

    private suspend fun settle() {
        if (latencyEnabled) delay(LATENCY_MILLIS)
        if (failing) throw ApiError.Network()
    }

    companion object {
        private const val LATENCY_MILLIS = 250L

        val ALL: List<BranchDetail> =
            listOf(
                BranchDetail(
                    id = MockIds.BRANCH_NISANTASI,
                    tenantId = MockIds.TENANT_NISANTASI,
                    slug = "nisantasi",
                    name = "Nişantaşı",
                    phone = "+902122345678",
                    address = "Teşvikiye Cad. No: 12, Şişli",
                ),
                BranchDetail(
                    id = MockIds.BRANCH_BODRUM,
                    tenantId = MockIds.TENANT_NISANTASI,
                    slug = "bodrum",
                    name = "Bodrum",
                    address = "Neyzen Tevfik Cad. No: 44, Bodrum",
                ),
            )
    }
}

private fun Patch<String>.resolve(old: String?): String? =
    when (this) {
        Patch.Unchanged -> old
        Patch.Clear -> null
        is Patch.Set -> value
    }

File: klinara-ios/klinara-ios/Services/Branches/BranchesService.swift
import Foundation

/// Şube yönetimi uçları ("Şube ve Personel").
///
/// `GET branches` giriş akışında ``AuthService/branches()`` ile de çağrılıyor
/// ama orada dar ``BranchSummary`` yeterli. Yönetim ekranı tam şekle bakıyor.
protocol BranchesService: Sendable {
    /// `GET /branches` — pasifler dahil kiracının tüm şubeleri.
    func branches() async throws -> [BranchDetail]
    /// `POST /branches`
    func create(_ input: CreateBranchInput) async throws -> BranchDetail
    /// `PATCH /branches/:id`
    func update(id: String, _ input: UpdateBranchInput) async throws -> BranchDetail
}

struct LiveBranchesService: BranchesService {

    private let client: APIClient

    init(client: APIClient) {
        self.client = client
    }

    func branches() async throws -> [BranchDetail] {
        let response: ListEnvelope<BranchDetail> = try await client.send(APIRequest.get("branches"))
        return response.data
    }

    func create(_ input: CreateBranchInput) async throws -> BranchDetail {
        try await client.send(APIRequest.post("branches", body: input))
    }

    func update(id: String, _ input: UpdateBranchInput) async throws -> BranchDetail {
        try await client.send(APIRequest.patch("branches/\(id)", body: input))
    }
}

/// Bellek-içi şube deposu. ``MockIDs`` şubeleriyle tohumlanıyor ki personel
/// ve takvim mock'larının şube kimlikleri burada da tanınsın.
final class MockBranchesService: BranchesService, @unchecked Sendable {

    private let lock = NSLock()
    private var _branches: [BranchDetail]

    init() {
        let created = Date(timeIntervalSince1970: 1_775_000_000)
        _branches = [
            BranchDetail(
                id: MockIDs.branchNisantasi, tenantId: MockIDs.tenant, slug: "nisantasi",
                name: "Nişantaşı", timezone: "Europe/Istanbul", phone: "+902122345678",
                address: "Teşvikiye Cad. No: 12, Şişli", isActive: true, createdAt: created
            ),
            BranchDetail(
                id: MockIDs.branchBagdat, tenantId: MockIDs.tenant, slug: "bagdat-caddesi",
                name: "Bağdat Caddesi", timezone: "Europe/Istanbul", phone: nil,
                address: "Bağdat Cad. No: 310, Kadıköy", isActive: true, createdAt: created
            ),
        ]
    }

    private func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock.lock()
        defer { lock.unlock() }
        return try body()
    }

    func branches() async throws -> [BranchDetail] {
        try? await Task.sleep(for: .seconds(0.3))
        return withLock { _branches }
    }

    func create(_ input: CreateBranchInput) async throws -> BranchDetail {
        try? await Task.sleep(for: .seconds(0.4))
        return try withLock {
            guard !_branches.contains(where: { $0.slug == input.slug }) else {
                throw APIError.problem(ProblemDetails(
                    code: .conflict,
                    title: "Bu şube kodu zaten kullanımda",
                    detail: "\"\(input.slug)\" bu klinikte başka bir şubeye ait.",
                    status: 409
                ))
            }
            let branch = BranchDetail(
                id: MockIDs.uuid(), tenantId: MockIDs.tenant, slug: input.slug, name: input.name,
                timezone: input.timezone ?? "Europe/Istanbul", phone: input.phone,
                address: input.address, isActive: true, createdAt: Date()
            )
            _branches.append(branch)
            return branch
        }
    }

    func update(id: String, _ input: UpdateBranchInput) async throws -> BranchDetail {
        try? await Task.sleep(for: .seconds(0.4))
        return try withLock {
            guard let index = _branches.firstIndex(where: { $0.id == id }) else {
                throw APIError.problem(ProblemDetails(code: .notFound, title: "Bulunamadı", status: 404))
            }
            let old = _branches[index]
            let updated = BranchDetail(
                id: old.id, tenantId: old.tenantId, slug: old.slug,
                name: input.name ?? old.name,
                timezone: input.timezone ?? old.timezone,
                phone: input.phone.applied(to: old.phone),
                address: input.address.applied(to: old.address),
                isActive: input.isActive ?? old.isActive,
                createdAt: old.createdAt
            )
            _branches[index] = updated
            return updated
        }
    }
}
